openfe/protocols/restraint_utils/geometry/boresch/guest.py

```diff
@@ -0,0 +1,251 @@
+
+    Returns
+    -------
+    list[tuple[int, int, int]]
+      A list of tuples containing all the angles.
+
+    Notes
+    -----
+    * In the original SepTop code at3 is picked as directly bonded to at1.
+      By comparison here we instead follow the case that at3 is bonded to
+      at2 but not bonded to at1.
+    """
+    angles = []
+
+    # Get the base atom and its neighbors
+    at1 = rdmol.GetAtomWithIdx(atom_idx)
+    at1_neighbors = [at.GetIdx() for at in at1.GetNeighbors()]
+
+    # We loop at2 and at3 through the sorted atom_pool in order to get
+    # a list of angles in the branch that are sorted by how close the atoms
+    # are from the central atom
+    for at2 in atom_pool:
+        if at2 in at1_neighbors:
+            at2_neighbors = [
+                at.GetIdx() for at in rdmol.GetAtomWithIdx(at2).GetNeighbors()
+            ]
+            for at3 in atom_pool:
+                if at3 != atom_idx and at3 in at2_neighbors:
+                    angles.append((atom_idx, at2, at3))
+
+    if aromatic_only:  # TODO: move this to its own method?
+        aromatic_rings = get_aromatic_rings(rdmol)
+
+        def _belongs_to_ring(angle, aromatic_rings):
+            for ring in aromatic_rings:
+                if all(a in ring for a in angle):
+                    return True
+            return False
+
+        for angle in angles:
+            if not _belongs_to_ring(angle, aromatic_rings):
+                angles.remove(angle)
+
+    return angles
+
+
+def _get_guest_atom_pool(
+    rdmol: Chem.Mol,
+    rmsf: npt.NDArray,
+    rmsf_cutoff: unit.Quantity
+) -> tuple[Optional[set[int]], bool]:
+    """
+    Filter atoms based on rmsf & rings, defaulting to heavy atoms if
+    there are not enough.
+
+    Parameters
+    ----------
+    rdmol : Chem.Mol
+      The RDKit Molecule to search through
+    rmsf : npt.NDArray
+      A 1-D array of RMSF values for each atom.
+    rmsf_cutoff : unit.Quantity
+      The rmsf cutoff value for selecting atoms in units compatible with
+      nanometer.
+
+    Returns
+    -------
+    atom_pool : Optional[set[int]]
+      A pool of candidate atoms.
+    ring_atoms_only : bool
+      True if only ring atoms were selected.
+    """
+    # Get a list of all the aromatic rings
+    # Note: no need to keep track of rings because we'll filter by
+    # bonded terms after, so if we only keep rings then all the bonded
+    # atoms should be within the same ring system.
+    atom_pool: set[tuple[int]] = set()
+    ring_atoms_only: bool = True
+    for ring in get_aromatic_rings(rdmol):
+        max_rmsf = rmsf[list(ring)].max()
+        if max_rmsf < rmsf_cutoff:
+            atom_pool.update(ring)
+
+    # if we don't have enough atoms just get all the heavy atoms
+    if len(atom_pool) < 3:
+        ring_atoms_only = False
+        heavy_atoms = get_heavy_atom_idxs(rdmol)
+        atom_pool = set(heavy_atoms[rmsf[heavy_atoms] < rmsf_cutoff])
```

Comment:
This is currently failing. Something like this could work:
```suggestion
        atom_pool = set(np.array(heavy_atoms)[np.array(rmsf[heavy_atoms] < rmsf_cutoff)])
```

Reply:
Thanks, there's something else odd in this code `atom_pool` is `set[tuple[int]]` which is nonsensical..

Reply:
This returns an array of `np.int64` which leads to an error in line 55 (`pos = conformer.GetAtomPosition(idx)`) which is expecting an `int.`
Maybe something like
```suggestion
        atom_pool = set(heavy_atoms[rmsf[heavy_atoms] < rmsf_cutoff].tolist())
```

Reply:
I'll think of something here, but generally it's better not to do these types of list operations, they get very expensive.